src/json.h

```diff
@@ -261,14 +288,40 @@ class JsonIn
 
         template<typename T, std::enable_if_t<std::is_enum<T>::value, int> = 0>
         bool read( T &val ) {
-            std::string s;
-            if( !read( s ) ) {
+            if( test_string() ) {
```

Comment:
I wonder: why can't you use the same logic as for reading `int` (no `test_numer`):
```C++
string s;
if( read(s ) ) {
    // handle string
}
int i;
if( read(i)) {
    // handle integer
}
error( ... )
```

`JsonIn::read( std::string & )` contains a `test_string` check, so you don't have to do this before calling it manually.

Reply:
I was trying to avoid constructing a `std::string` unnecessarily, but you're probably right; that's not a real concern.

Reply:
Than try to load it as int first. If that fails, try to load as string. If that fails, error out.

The only time the string has been constructed *unnecessarily* is within the failure path and there it won't matter as propagating and displaying the error is costly enough and not performance critical anyway.